ThunderCloud/TSCAppDelegate.swift

```diff
@@ -0,0 +1,196 @@
+//
+//  TSCAppDelegate.swift
+//  ThunderCloud
+//
+//  Created by Simon Mitchell on 24/08/2017.
+//  Copyright © 2017 threesidedcube. All rights reserved.
+//
+
+import UIKit
+import UserNotifications
+import ThunderRequest
+
+@UIApplicationMain
+/// A root app delegate which sets up your window and push notifications e.t.c.
+open class TSCAppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate {
+
+	/// The main window of the app
+	public var window: UIWindow?
+	
+	/// Whether to show push notifications when the app is in the foreground
+	public var foregroundNotificationOptions: UNNotificationPresentationOptions? = [.alert, .badge, .sound]
+	
+	open func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
+		
+		UNUserNotificationCenter.current().delegate = self
+		
+		window = UIWindow(frame: UIScreen.main.bounds)
+		window?.backgroundColor = .white
+		
+		window?.rootViewController = TSCAppViewController()
+		window?.makeKeyAndVisible()
+		
+		setupSharedUserAgent()
+		
+		DeveloperModeController.shared.installDeveloperMode(toWindow: window!, currentTheme: TSCTheme())
+		
+		// Register errors
+		TSCErrorRecoveryAttempter.registerOverrideDescription(
+			"Failed to load page".localised(with: "_STREAMINGPAGE_FAILED_TITLE"),
+			recoverySuggestion: "We were unable to find the page for this notification".localised(with: "_STREAMINGPAGE_FAILED_RECOVERYSUGGESTION"),
+			forDomain: "ThunderCloud.streamingError",
+			code: 1)
+				
+		return true
+	}
+	
+	//MARK: -
+	//MARK: - Push notifications
+	//MARK: -
+	
+	open func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+		TSCStormNotificationHelper.registerPushToken(deviceToken)
+	}
+
+	open func handleNotificationResponse(for notification: UNNotification, response: UNNotificationResponse?, fromLaunch: Bool) -> Bool {
+		
+		// Make sure the action wasn't dismissing the notification
+		guard response?.actionIdentifier != UNNotificationDismissActionIdentifier else {
+			return false
+		}
+		
+		guard let payload = notification.request.content.userInfo["payload"] as? [AnyHashable : Any] else { return false }
+		guard let url = payload["url"] as? String else { return false }
+		
+		// Call handleNotificationResponse(for:response:fromLaunch) every second if we are presenting
+		if window?.rootViewController?.presentedViewController != nil {
+			
+			DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+				_ = self.handleNotificationResponse(for: notification, response: response, fromLaunch: fromLaunch)
+			}
+			
+		} else {
+			
+			return handleContentNotificationFor(cacheURL: url)
+		}
+		
+		return true
+	}
+	
+	open func handleContentNotificationFor(cacheURL: String) -> Bool {
+		
+		if Bundle.main.infoDictionary?["TSCStreamingBaseURL"] as? String != nil {
```

Comment:
I'm unsure about what could be done here, whether perhaps each part of this if-statement should in fact be moved out to two separate functions, which get the `cacheURL` passed into them. That way, it reduces the level of indentation, especially as there's blocks in use, and helps with self-documenting the code (and documentation is pretty awesome!) Even further, there looks like there may be some code that can be reused across both flows of the statement.  What are your opinions? :)

Reply:
I'd like to keep it as-is for now. It's unlikely we'll be testing this because it's heavily UI based and reliant on network usage which isn't easy to do in test suites AFAIK. But mainly I'd like to keep it as-is because I want to keep the App Delegate as clean as possible, with as few functions as possible. Ideally the App Delegate should only have delegate methods for `UIApplicationDelegate` in it, we need to investigate something @MattCheetham shared with us a while back to try and clean up the `AppDelegate` a bit, so I'm okay with this being as is for a while 😊